### ISSUE TITLE
Parsing to data class with many values gives random result

### ISSUE THREAD

**Issue by avently** [Kotlin/kotlinx.serialization#377]
Found sooo strange behaviour. I have data class with ~100 values in it. When I'm parsing a string (`json.parse(class.serializer(), someString)`) or string -> JsonArray -> JsonElement -> data class (`json.fromJson(class.serializer(), jsonElement`)) values in data class are filled randomly. Some of them located on their places but some of them missing. So, take a look at the code:
```kotlin
val s =  """{"value10":"Anything","value11":3456.7,"value33":"SomeString","value26":0.123,"value38":"2019-02-15T07:58:41.519Z"}"""
        println("SHOULD BE THIS: $s")
        println("BUT GOT THIS: " + json.parse(Testing.serializer(), s))


@Serializable
data class Testing (
        @Optional
        val value1: String? = null,
        @Optional
        val value2: String? = null,
        @Optional
        val value3: String? = null,
        @Optional
        val value4: String? = null,
        @Optional
        val value5: String? = null,
        @Optional
        val value6: Double? = null,
        @Optional
        val value7: Double? = null,
        @Optional
        val value8: Double? = null,
        @Optional
        val value9: Double? = null,
        @Optional
        val value10: String? = null,
        @Optional
        val value11: Double? = null,
        @Optional
        val value12: Double? = null,
        @Optional
        val value13: Double? = null,
        @Optional
        val value14: Boolean = false,
        @Optional
        val value15: Boolean = false,
        @Optional
        val value16: Double? = null,
        @Optional
        val value17: Double? = null,
        @Optional
        val value18: Double? = null,
        @Optional
        val value19: Double? = null,
        @Optional
        val value20: Double? = null,
        @Optional
        val value21: Double? = null,
        @Optional
        val value22: Double? = null,
        @Optional
        val value23: Long? = null,
        @Optional
        val value24: Long? = null,
        @Optional
        val value25: Long? = null,
        @Optional
        val value26: Double? = null,
        @Optional
        val value27: Double? = null,
        @Optional
        val value28: Double? = null,
        @Optional
        val value29: Double? = null,
        @Optional
        val value30: Double? = null,
        @Optional
        val value31: Double? = null,
        @Optional
        val value32: Double? = null,
        @Optional
        val value33: String? = null,
        @Optional
        val value34: Double? = null,
        @Optional
        val value35: Double? = null,
        @Optional
        val value36: Double? = null,
        @Optional
        val value37: Double? = null,
        @Optional
        val value38: String? = null)
```

```
SHOULD BE THIS: {"value10":"Anything","value11":3456.7,"value33":"SomeString","value26":0.123,"value38":"2019-02-15T07:58:41.519Z"}
    BUT GOT THIS: Testing(value1=null, value2=null, value3=null, value4=null, value5=null, value6=null, value7=null, value8=null, value9=null, value10=Anything, value11=3456.7, value12=null, value13=null, value14=false, value15=false, value16=null, value17=null, value18=null, value19=null, value20=null, value21=null, value22=null, value23=null, value24=null, value25=null, value26=0.123, value27=null, value28=null, value29=null, value30=null, value31=null, value32=null, value33=null, value34=null, value35=null, value36=null, value37=null, value38=null)
```

Now look at value33 and value38. They are null but shouldn't be. Other values on their places.
So it happens really randomly. In my situation application retrieves data via web socket and some values are present right now but in next second other values will be present. And some values after parsing are missing. Its not the same every time. Is it a bug? Or feature?:)

**Comment by sandwwraith** [comment]
It is a known issue (#164) which will likely be fixed soon

**Comment by avently** [comment]
@sandwwraith are you working on it? Issue already known since last July so how you can expect "to be fixed soon".

**Comment by avently** [comment]
Ok, new information.
I looked at the compiled code from data class.
Java (or Kotlin?) writes something like:
```java
// $FF: synthetic method
   public Tester(String var1, String var2, String var3, String var4, String var5, String var6, String var7, String var8, String var9, String var10, String var11, String var12, String var13, String var14, String var15, String var16, String var17, String var18, String var19, String var20, String var21, BigDecimal var22, BigDecimal var23, BigDecimal var24, BigDecimal var25, BigDecimal var26, String var27, BigDecimal var28, BigDecimal var29, BigDecimal var30, boolean var31, boolean var32, BigDecimal var33, BigDecimal var34, Long var35, Long var36, BigDecimal var37, Boolean var38, Boolean var39, Boolean var40, BigDecimal var41, BigDecimal var42, BigDecimal var43, BigDecimal var44, String var45, String var46, String var47, String var48, String var49, BigDecimal var50, BigDecimal var51, String var52, String var53, String var54, String var55, String var56, BigDecimal var57, BigDecimal var58, BigDecimal var59, BigDecimal var60, BigDecimal var61, BigDecimal var62, BigDecimal var63, BigDecimal var64, BigDecimal var65, Long var66, Long var67, Long var68, BigDecimal var69, BigDecimal var70, BigDecimal var71, BigDecimal var72, BigDecimal var73, BigDecimal var74, BigDecimal var75, String var76, BigDecimal var77, BigDecimal var78, BigDecimal var79, BigDecimal var80, BigDecimal var81, BigDecimal var82, BigDecimal var83, boolean var84, BigDecimal var85, BigDecimal var86, String var87, BigDecimal var88, BigDecimal var89, BigDecimal var90, String var91, BigDecimal var92, BigDecimal var93, BigDecimal var94, BigDecimal var95, String var96, int var97, int var98, int var99, DefaultConstructorMarker var100) {

      if ((var97 & 1) != 0) {
         var1 = (String)null;
      }

      if ((var97 & 2) != 0) {
         var2 = (String)null;
      }

      if ((var97 & 4) != 0) {
         var3 = var0.state;
      }

      if ((var97 & 8) != 0) {
         var4 = var0.typ;
      }

      if ((var97 & 16) != 0) {
         var5 = var0.listing;
      }

      if ((var97 & 32) != 0) {
         var6 = var0.front;
      }
.............

     if ((var97 & 1073741824) != 0) {
         var31 = false;
      }

      if ((var97 & Integer.MIN_VALUE) != 0) {
         var32 = false;
      }

      if ((var98 & 1) != 0) {
         var33 = (BigDecimal)null;
      }

      if ((var98 & 2) != 0) {
         var34 = (BigDecimal)null;
      }
.........

      if ((var98 & 1073741824) != 0) {
         var63 = var0.totalVolume;
      }

      if ((var98 & Integer.MIN_VALUE) != 0) {
         var64 = var0.volume;
      }

      if ((var99 & 1) != 0) {
         var65 = var0.volume24h;
      }

      if ((var99 & 2) != 0) {
         var66 = var0.prevTotalTurnover;
      }

..............

      if ((var99 & 1073741824) != 0) {
         var95 = var0.settledPrice;
      }

      if ((var99 & Integer.MIN_VALUE) != 0) {
         var96 = var0.timestamp;
      }
}
```

So as you can see, code makes bitwise operations and choose what to do based on it. Interesting thing here that var97, var98, var99 used for that bitwise operation. There are 3 numbers because Integer.MAX_VALUE is reached after 32 operations. So var98 is used after first 32 operations. 

Now take a look at a code generated by Gradle serialization plugin:

```java
public Tester(int seen, @Optional @Nullable String value1, @Optional @Nullable String value2, @Optional @Nullable String value3, @Optional @Nullable String value4, @Optional @Nullable String value5, @Optional @Nullable String value6, @Optional @Nullable String value7, @Optional @Nullable String value8, @Optional @Nullable String value9, @Optional @Nullable String value10, @Optional @Nullable String value11, @Optional @Nullable String value12, @Optional @Nullable String value13, @Optional @Nullable String value14  ............., @Nullable SerializationConstructorMarker serializationConstructorMarker) {
      super();
      if ((seen & 1) != 0) {
         this.value1 = value1;
      } else {
         this.value1 = null;
      }

      if ((seen & 2) != 0) {
         this.value2 = value2;
      } else {
         this.value2 = null;
      }

      if ((seen & 4) != 0) {
         this.value3 = value3;
      } else {
         this.value3 = null;
      }
     ...............
     
     if ((seen & 1073741824) != 0) {
         this.value31 = value31;
      } else {
         this.value31 = false;
      }

      if ((seen & Integer.MIN_VALUE) != 0) {
         this.value32 = value32;
      } else {
         this.value32 = false;
      }

// LOOK HERE:

      if ((seen & 1) != 0) {
         this.value33 = value33;
      } else {
         this.value33 = null;
      }

      if ((seen & 2) != 0) {
         this.value34 = value34;
      } else {
         this.value34 = null;
      }

```

As you can see after my comment `seen` variable is used over and over in loop. That's why data class has incorrect data in it after 32 element. 

Let's go to the next method:

```java
@NotNull
      public Tester deserialize(@NotNull Decoder decoder) {
Intrinsics.checkParameterIsNotNull(decoder, "decoder");
         SerialDescriptor var2 = $$serialDesc;
         boolean var4 = false;
         int var5 = 0;
         int var6 = 0;
         int var7 = 0;
         boolean var8 = false;
         String var9 = null;
         String var10 = null;
         String var11 = null;
         String var12 = null;
         String var13 = null;
................

case 94:
               var10003 = (KSerializer)BigDecimalSerializer.INSTANCE;
               var103 = (BigDecimal)((var7 & 1073741824) != 0 ? decoder.updateNullableSerializableElement(var2, 94, var10003, var103) : decoder.decodeNullableSerializableElement(var2, 94, var10003));
               var7 |= 1073741824;
               if (!var4) {
                  break;
               }
            case 95:
               var10003 = (KSerializer)StringSerializer.INSTANCE;
               var104 = (String)((var7 & Integer.MIN_VALUE) != 0 ? decoder.updateNullableSerializableElement(var2, 95, var10003, var104) : decoder.decodeNullableSerializableElement(var2, 95, var10003));
               var7 |= Integer.MIN_VALUE;
               if (!var4) {
                  break;
               }
            case -1:
               decoder.endStructure(var2);
               return new Tester(var5, var9, var10, var11, var12, var13, var14, var15, var16, var17, var18, var19, var20, var21, var22, var23, var24, var25, var26, var27, var28, var29, var30, var31, var32, var33, var34, var35, var36, var37, var38, var39, var40, var41, var42, var43, var44, var45, var46, var47, var48, var49, var50, var51, var52, var53, var54, var55, var56, var57, var58, var59, var60, var61, var62, var63, var64, var65, var66, var67, var68, var69, var70, var71, var72, var73, var74, var75, var76, var77, var78, var79, var80, var81, var82, var83, var84, var85, var86, var87, var88, var89, var90, var91, var92, var93, var94, var95, var96, var97, var98, var99, var100, var101, var102, var103, var104, (SerializationConstructorMarker)null);
            default:
               throw (Throwable)(new UnknownFieldException(var3));
            }

```

Here method works right and contains three variables that was used for calculations: var5, var6, var7. But in the end of the method it calls:
```java
 return new Tester(var5, var9, var10, var11, var12, var13
```
with only one variable var5 (which goes to `seen` var as we already saw). 
So fix should be pretty simple. Hope it helps.

**Comment by q-litzler** [comment]
I'm having the same issue, it would help tremendously if this could be fixed soon ! Is there any workaround in the meantime ? Thanks a lot

**Comment by sandwwraith** [comment]
This is gonna be fixed in the next Kotlin release (since it is a problem in plugin which is shipped with compiler)

**Comment by q-litzler** [comment]
> This is gonna be fixed in the next Kotlin release (since it is a problem in plugin which is shipped with compiler)

Ok, thank you !

**Comment by avently** [comment]
Yep, thank you, @sandwwraith , as always

**Comment by Amagi82** [comment]
@sandwwraith Do you have a link to the related plugin issue on the bug tracker?

**Comment by zhewang2** [comment]
We have same the issue [#442](https://github.com/Kotlin/kotlinx.serialization/issues/422).

Do you have any idea when the next Kotlin release that has the fix for this bug is?

**Comment by sandwwraith** [comment]
@zhewang2 Kotlin 1.3.30 will have the fix. Can I close #442 as duplicate?

**Comment by zhewang2** [comment]
@sandwwraith sure. Thank you for letting us know! Cannot wait to use it!

**Comment by sandwwraith** [comment]
Fixed with #164 in 1.3.30
